resources/aws/securitygroup.go

```diff
@@ -34,30 +50,51 @@ func (s SecurityGroup) findExisting() (*ec2.SecurityGroup, error) {
 				},
 			},
 		},
-	})
-	if err != nil {
-		return nil, microerror.MaskAny(err)
 	}
 
-	if len(securityGroups.SecurityGroups) < 1 {
-		return nil, microerror.MaskAny(securityGroupFindError)
+	for _, filter := range filters {
+		securityGroups, err := s.Clients.EC2.DescribeSecurityGroups(&ec2.DescribeSecurityGroupsInput{
+			Filters: filter,
+		})
+		if err != nil {
+			return nil, microerror.MaskAny(err)
+		}
+
+		if len(securityGroups.SecurityGroups) < 1 {
+			continue
+		}
+
+		return securityGroups.SecurityGroups[0], nil
 	}
 
-	return securityGroups.SecurityGroups[0], nil
+	return nil, microerror.MaskAny(securityGroupFindError)
 }
 
-func (s *SecurityGroup) CreateIfNotExists() (bool, error) {
-	if err := s.CreateOrFail(); err != nil {
-		if strings.Contains(err.Error(), awsclient.SecurityGroupDuplicate) {
-			securityGroup, err := s.findExisting()
-			if err != nil {
-				return false, microerror.MaskAny(err)
-			}
-			s.id = *securityGroup.GroupId
-
+func (s *SecurityGroup) checkIfExists() (bool, error) {
+	securityGroup, err := s.findExisting()
+	if err != nil {
+		if IsSecurityGroupFindError(err) {
 			return false, nil
 		}
+		return false, microerror.MaskAny(err)
+	}
+
+	s.id = *securityGroup.GroupId
```

Comment:
I see, but that does not mean it is right. I find it wrong. The methods name and interface suggest it only checks for some boolean. Nobody will ether guess that internal state is manipulated. I think it would be good to create an issue to fix this in general. IMO the design of the function or even the overall business logic of how to retrieve information is than weird.